server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.readiness;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.network.NetworkAddress;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.transport.BoundTransportAddress;
+import org.elasticsearch.common.transport.PortsRange;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.shutdown.PluginShutdownService;
+import org.elasticsearch.transport.BindTransportException;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.StandardProtocolFamily;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static org.elasticsearch.node.Node.WRITE_PORTS_FILE_SETTING;
+import static org.elasticsearch.node.Node.writePortsFile;
+
+public class ReadinessService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(ReadinessService.class);
+
+    private final Environment environment;
+
+    private volatile boolean active = false;
+    private volatile ServerSocketChannel serverChannel;
+    final AtomicReference<InetSocketAddress> boundSocket = new AtomicReference<>();
+
+    public static final Setting<String> PORT = new Setting<>(
+        "readiness.port",
+        "9400-9500",
```

Comment:
Copying from https://github.com/elastic/elasticsearch/issues/81168#issuecomment-1063332832: we think that scanning for a free port is trappy (https://github.com/elastic/elasticsearch/issues/54354), could we avoid that and instead only accept a single port here?

Reply:
OK, sounds good. I'll ensure that this is configured explicitly and only enable the service if it's configured.